src/connector/src/error.rs

```diff
@@ -12,50 +12,15 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use thiserror::Error;
-
-#[derive(Error, Debug)]
-pub enum ConnectorError {
-    #[error("Parse error: {0}")]
-    Parse(&'static str),
-
-    #[error("Invalid parameter {name}: {reason}")]
-    InvalidParam { name: &'static str, reason: String },
-
-    #[error("Kafka error: {0}")]
-    Kafka(#[from] rdkafka::error::KafkaError),
-
-    #[error("Config error: {0}")]
-    Config(
-        #[source]
-        #[backtrace]
-        anyhow::Error,
-    ),
-
-    #[error("Connection error: {0}")]
-    Connection(
-        #[source]
-        #[backtrace]
-        anyhow::Error,
-    ),
-
-    #[error("MySQL error: {0}")]
-    MySql(#[from] mysql_async::Error),
```

Comment:
> anyhow::Error is good enough if one just wants to make the error type informative but not necessarily actionable. However, given a value of type anyhow::Error, it is hard to tell which module or crate it comes from, which may blur the boundary between modules when passing it around, leading to abuse.

I think the new `ConnectorError` on the right side will lose the context of specific connector, e.g. `MySQL error`. So it is just a new type of the `anyhow::Error`, and doesn't embed information of specific crate/module. 

Without the context information, we don't know which module throws the error and must look into the stacktrace to find out. For example, previously we can tell the IO error comes from mysql https://github.com/risingwavelabs/risingwave/issues/14847.

So I -1 for current implementation, could you embed the crate/module info into the macro?

Reply:
Specifically for #14847, I believe the retrying should be done **before** throwing the `mysql_async::Error` to upper layer, which is, before turning it into a `ConnectorError`. This is because performing ad-hoc matching for MySQL in the general code path of handling type-erased `ConnectorError` appears to be an abstraction leak. In this case, the internal structure of `ConnectorError` is not relevant.

However, if we're intend to do that on `ConnectorError`, `anyhow` still allows developer to downcast to a concrete type, just like trait objects.

Reply:
There's a principle that errors should be either informative or actionable. In contrast of implementing the retry logic (actionable), I suppose the more reasonable part is that the refactor will lose some context in the error **message** (informative) and make it vaguer, as discussed in https://github.com/risingwavelabs/risingwave/pull/15031/files#r1479489349.

Specifically, the message will become:

```diff
- Connector error: MySQL error: Input/output error: Input/output error: can't parse: buf doesn't have enough data
+ Connector error: Input/output error: Input/output error: can't parse: buf doesn't have enough data
```

I admit that this undoubtedly make it less informative and harder to identify the root cause of external services at a glance. This could be mainly because the error messages from 3rd-party crates are not managed and reviewed by us. As a result, we are less familiar with them and could confuse them with others.

----

I'm okay to keep the original `enum` implementation. As pointed out by @xxchan, a prefix of `MySQL error: ` is not as good as a manually-specified context message like `Failed to read the offset from MySQL: `, but could be still better than no context. We could improve that in the future.

Reply:
Yeah, I am most concerning the losing context problem which would increase the burden of troubleshooting. Ideally, if we can extract 3rd party crate name into the `def_anyhow_newtype` macro, then we can abandon the original `enum` implementation.

Reply:
Based on the discussion with @xxchan, I've come up with another style of `anyhow` wrapper which will force the developers to provide contexts or detailed messages when upcasting the error. Will explore it in the next PRs.